sys/net/gnrc/routing/ipv6_auto_subnets/gnrc_ipv6_auto_subnets.c

```diff
@@ -52,14 +61,121 @@
  * @author  Benjamin Valentin <benjamin.valentin@ml-pa.com>
  */
 
+#include "net/gnrc/ipv6.h"
+#include "net/gnrc/netif.h"
+#include "net/gnrc/netif/hdr.h"
+#include "net/gnrc/udp.h"
 #include "net/gnrc/ipv6/nib.h"
 #include "net/gnrc/ndp.h"
+#include "random.h"
+
+/**
+ * @brief Port for the custom UDP sync protocol
+ */
+#ifndef CONFIG_GNRC_IPV6_AUTO_SUBNETS_PORT
+#define CONFIG_GNRC_IPV6_AUTO_SUBNETS_PORT          (16179)
+#endif
+
+/**
+ * @brief Max number of other routers on the same link
+ */
+#ifndef CONFIG_GNRC_IPV6_AUTO_SUBNETS_PEERS_MAX
+#define CONFIG_GNRC_IPV6_AUTO_SUBNETS_PEERS_MAX     (4)
+#endif
+
+/**
+ * @brief How often the number subnets should be announced by the routers
+ */
+#ifndef CONFIG_GNRC_IPV6_AUTO_SUBNETS_TX_RETRIES
+#define CONFIG_GNRC_IPV6_AUTO_SUBNETS_TX_RETRIES    (3)
+#endif
+/**
+ * @brief How long to wait for other routers annoucements before resending
+ *        or creating subnets when the retry counter is exhausted
+ */
+#ifndef CONFIG_GNRC_IPV6_AUTO_SUBNETS_TIMEOUT_MS
+#define CONFIG_GNRC_IPV6_AUTO_SUBNETS_TIMEOUT_MS    (50)
+#endif
+
+#define SERVER_THREAD_STACKSIZE                     (THREAD_STACKSIZE_DEFAULT)
+#define SERVER_MSG_QUEUE_SIZE                       (CONFIG_GNRC_IPV6_AUTO_SUBNETS_PEERS_MAX)
+#define SERVER_MSG_TYPE_TIMEOUT                     (0x8fae)
 
 #define ENABLE_DEBUG 0
 #include "debug.h"
 
 static char addr_str[IPV6_ADDR_MAX_STR_LEN];
 
+#if !IS_USED(MODULE_GNRC_IPV6_AUTO_SUBNETS_SIMPLE)
+
+/**
+ * @brief Custom UDP sync protocol
+ */
+typedef struct __attribute__((packed)) {
+    uint8_t num_subnets;    /**< number of subnets a host wants to create */
+} auto_subnets_request_t;
+
+/* keep a copy of PIO information in memory */
+static gnrc_netif_t *_upstream;
+static ndp_opt_pi_t _pio_cache;
+
+static char auto_subnets_stack[SERVER_THREAD_STACKSIZE];
+static msg_t server_queue[SERVER_MSG_QUEUE_SIZE];
+
+/* store neighbor routers l2 address to ignore duplicate packets */
+static uint8_t l2addrs[CONFIG_GNRC_IPV6_AUTO_SUBNETS_PEERS_MAX]
+                      [CONFIG_GNRC_IPV6_NIB_L2ADDR_MAX_LEN];
```

Comment:
We could get rid of this array by instead adding a flag to the entries in the NIB. This would however require to mess with nib internals that are not accessible through the public API, so I decided to rather keep a local collection of neighboring routers l2 addressees here.

Reply:
Couldn't you re-use something from your neighborhood discovery stuff you added for MRHOF?

Reply:
Other than that, I think its fine :-).

Reply:
Do you mean `netstats_neighbor`? That would be unnecessary overhead since it stores *all* neighbors, not just routers. And since this is intended for wired networks, it's unlikely `netstats_neighbor` is otherwise useful.

Reply:
Ok